### PR TITLE
SpecOp bool->uint/int and uint<->int conversion

### PR DESCRIPTION
**Pull request by Qining**
Bool -> uint/int with OpSpecConstantOp OpSelect instruction.

uint <-> int conversion with OpSpecConstantOp OpIAdd instruction.

Note, implicit conversion: `const uint = an_int_spec_constant` is not
supported. Explicit type casting is required: `const uint =
uint(an_int_spec_constant)`

### CODE REVIEW

File: SPIRV/GlslangToSpv.cpp [hunk 1]
```diff
@@ -3307,7 +3306,27 @@ spv::Id TGlslangToSpvTraverser::createConversion(glslang::TOperator op, spv::Dec
...
+        }
+        // Don't 'break' here as this case should be grouped together with
+        // EOpConvIntToUint when generating normal run-time conversion
+        // instruction.
     case glslang::EOpConvIntToUint:
+        if (builder.isInSpecConstCodeGenMode()) {
+            // Build zero scalar or vector for OpIAdd.
+            if (zero == 0) {
+                zero = builder.makeUintConstant(0);
+                zero = makeSmearedConstant(zero, vectorSize);
```
Comment (Qining) [inline SPIRV/GlslangToSpv.cpp]: Is there a way to differentiate unsigned int type and signed int type? Such a function can make this part much nicer.

Comment (johnkslang) [inline SPIRV/GlslangToSpv.cpp]: Not sure I understand.  Yes, we differentiate, and you chose unsigned, and so ```zero```'s type is unsigned, which is probably the best choice.  You could have chosen a signed-int type for 0, and that would also work, but unsigned is more appropriate.

File: SPIRV/GlslangToSpv.cpp [hunk 2]
```diff
@@ -3307,7 +3306,27 @@ spv::Id TGlslangToSpvTraverser::createConversion(glslang::TOperator op, spv::Dec
         break;
 
     case glslang::EOpConvUintToInt:
+        if (builder.isInSpecConstCodeGenMode()) {
+            // Build zero scalar or vector for OpIAdd to do the conversion when
+            // generating for OpSpecConstantOp instruction.
+            zero = builder.makeIntConstant(0);
```
Comment (johnkslang) [inline SPIRV/GlslangToSpv.cpp]: I think you can always fall through to do this, and just have one piece of code that uses an unsigned type for `zero`.

File: SPIRV/SpvBuilder.cpp [hunk 3]
```diff
@@ -1212,6 +1212,14 @@ Id Builder::createBinOp(Op opCode, Id typeId, Id left, Id right)
 
 Id Builder::createTriOp(Op opCode, Id typeId, Id op1, Id op2, Id op3)
 {
+    // Generate code for spec constants if in spec constant operation
+    // generation mode.
+    if (generatingOpCodeForSpecConst) {
+        std::vector<Id> operands(3);
+        operands[0] = op1; operands[1] = op2; operands[2] = op3;
```
Comment (johnkslang) [inline SPIRV/SpvBuilder.cpp]: Would like one statement per line, thanks.

File: SPIRV/GlslangToSpv.cpp (hunk 1 above)
Comment (Qining) [inline SPIRV/GlslangToSpv.cpp]: Hmm, actually I hope to let the EOpConvUintToInt case fall through. Just hope to have function to tell whether a type ID is uint type or int type. I found this not that easy, as both int and uint are declared with OpTypeInt. It seems like the only way to tell is to check the OpTypeInt instruction's first operand.

Comment (johnkslang) [inline SPIRV/GlslangToSpv.cpp]: You don't need to tell. Just always make uint for `zero`, or is it some other reason you want to know?

Comment (Qining) [inline SPIRV/GlslangToSpv.cpp]: Just think having it different for int and uint may looks a little bit better. Anyway, already changed it to use 'uint' zero for both.

**Pull request merged**